### PR TITLE
update tensorrt parser to support TRT 8.2

### PR DESCRIPTION
**Pull request by askhade**
Triton 21.12 will include TRT 8.2. This PR is in preparation for this change.

With this change we let the user to provide a branch\tag for onnx-tensorrt parser repo "TRITON_ONNX_TENSORRT_REPO_TAG "

If this tag is specified we simply use this to build ORT. If nothing is specified then we use the defaults listed in ORT_TO_TRTPARSER_VERSION_MAP specific to the ort and trt version being built. If this map does not have an entry for the ort version and trt version combo then we simply use the default branch\tag which ort uses.

This is to enable support for versions of TRT which ORT release does not support yet. 

Note: I have tested this on linux but not windows.

### CODE REVIEW

File: tools/gen_ort_dockerfile.py
Comment (jywu-msft) [inline tools/gen_ort_dockerfile.py:38]: I think we can wait for the GA version of the parser to merge this?
EA was primarily for us to start testing earlier. 
@kevinch-nv , I assume there will be an 8.2 ga branch created in the next week or two?

Comment (askhade) [reply tools/gen_ort_dockerfile.py:38]: Right! this PR is meant for us to start testing for 21.12 - Triton-ORT-TRT8.2

Comment (kevinch-nv) [reply tools/gen_ort_dockerfile.py:38]: Yes, the GA branch will be created once the 8.2 GA is pubicaly released.

**Comment by deadeyegoodwin** [comment]
I converted this PR to a draft so it doesn't accidentally get merged

File: CMakeLists.txt
Comment (deadeyegoodwin) [inline CMakeLists.txt:63]: "... to specify a branch in..."

File: tools/gen_ort_dockerfile.py
Comment (deadeyegoodwin) [inline tools/gen_ort_dockerfile.py:525]: corresponding

**Review by deadeyegoodwin — CHANGES_REQUESTED** [review]

File: tools/gen_ort_dockerfile.py
Comment (deadeyegoodwin) [inline tools/gen_ort_dockerfile.py:37]: Do we need the Triton container version? Can't we just have a map from ORT version to corresponding parser version?

**Review by deadeyegoodwin — CHANGES_REQUESTED** [review]

File: tools/gen_ort_dockerfile.py
Comment (askhade) [reply tools/gen_ort_dockerfile.py:37]: No it is not enough... we either need triton version or trt version... because which version of parser to pick depends on the trt version. I chose to use triton container version instead of trt version here because we already have the info regarding triton version so simply using it is easy.

Comment (deadeyegoodwin) [reply tools/gen_ort_dockerfile.py:37]: The problem is that linux x86, linux arm, jetson and windows may use different versions of TensorRT in each release. I think we need to extract the TRT version from the image in build.py and then pass it to ORT backend build by another ARG. Then your logic would be as it is now except using TRT version (if given) instead of Triton version. I can make the build.py changes to determine the TRT version.

**Review by deadeyegoodwin — CHANGES_REQUESTED** [review]

File: tools/gen_ort_dockerfile.py
Comment (askhade) [reply tools/gen_ort_dockerfile.py:37]: ok makes sense. I will update this PR to use trt version instead...

File: tools/gen_ort_dockerfile.py
Comment (deadeyegoodwin) [inline tools/gen_ort_dockerfile.py:224]: Does this PR need to be rebased? It seems to have the debug build changes. Can we merge the debug changes and rebase this? Or are debug changes not ready yet?

**Comment by deadeyegoodwin** [comment]
I'll working on updating build.py to provide TRITON_TENSORRT_VERSION to the onnx runtime build.

File: tools/gen_ort_dockerfile.py
Comment (askhade) [reply tools/gen_ort_dockerfile.py]: Rebase Done.

**Review by deadeyegoodwin — APPROVED** [review]

File: tools/gen_ort_dockerfile.py
Comment (deadeyegoodwin) [inline tools/gen_ort_dockerfile.py:37]: The TensorRT versions that you will see will be a.b.c.d or even something like 8.0.3.4+cuda11.3.1.005
So I think you will need to parse --trt-version to look only at a.b portion of the version

**Review by deadeyegoodwin — CHANGES_REQUESTED** [review]

**Comment by deadeyegoodwin** [comment]
@askhade 
server PR: https://github.com/triton-inference-server/server/pull/3588

Here is how onnxruntime_backend cmake will be invoked once PR is merged

cmake ['-DTRITON_BUILD_ONNXRUNTIME_VERSION=1.9.0', '-DTRITON_ENABLE_ONNXRUNTIME_TENSORRT=ON', '-DTRITON_BUILD_CONTAINER=gitlab-master.nvidia.com:5005/dl/dgx/cuda:11.4-devel-ubuntu20.04--21.10', '-DTRITON_ENABLE_ONNXRUNTIME_OPENVINO=ON', '-DTRITON_BUILD_ONNXRUNTIME_OPENVINO_VERSION=2021.2.200', '-DCMAKE_BUILD_TYPE=Release', '-DCMAKE_INSTALL_PREFIX:PATH=/tmp/tritonbuild/onnxruntime/install', '-DTRITON_COMMON_REPO_TAG:STRING=main', '-DTRITON_CORE_REPO_TAG:STRING=main', '-DTRITON_BACKEND_REPO_TAG:STRING=main', '-DTRITON_ENABLE_GPU:BOOL=ON', '-DTRITON_ENABLE_MALI_GPU:BOOL=OFF', '-DTRT_VERSION=8.0.3.4+cuda11.3.1.005', '-DDALI_VERSION=1.6.0', '..']

**Comment by askhade** [comment]
> @askhade server PR: [triton-inference-server/server#3588](https://github.com/triton-inference-server/server/pull/3588)
> 
> Here is how onnxruntime_backend cmake will be invoked once PR is merged
> 
> cmake ['-DTRITON_BUILD_ONNXRUNTIME_VERSION=1.9.0', '-DTRITON_ENABLE_ONNXRUNTIME_TENSORRT=ON', '-DTRITON_BUILD_CONTAINER=gitlab-master.nvidia.com:5005/dl/dgx/cuda:11.4-devel-ubuntu20.04--21.10', '-DTRITON_ENABLE_ONNXRUNTIME_OPENVINO=ON', '-DTRITON_BUILD_ONNXRUNTIME_OPENVINO_VERSION=2021.2.200', '-DCMAKE_BUILD_TYPE=Release', '-DCMAKE_INSTALL_PREFIX:PATH=/tmp/tritonbuild/onnxruntime/install', '-DTRITON_COMMON_REPO_TAG:STRING=main', '-DTRITON_CORE_REPO_TAG:STRING=main', '-DTRITON_BACKEND_REPO_TAG:STRING=main', '-DTRITON_ENABLE_GPU:BOOL=ON', '-DTRITON_ENABLE_MALI_GPU:BOOL=OFF', '-DTRT_VERSION=8.0.3.4+cuda11.3.1.005', '-DDALI_VERSION=1.6.0', '..']

Thanks @deadeyegoodwin . I will fix the PR to correctly pick up the version.

**Review by GuanLuo — APPROVED** [review]

**Review by deadeyegoodwin — APPROVED** [review]

**Pull request merged**
